### ISSUE TITLE
IDX fields in BCF2.2's VCF header

### ISSUE THREAD

**Issue by h-2** [samtools/hts-specs#591]
I am currently implementing a stand-alone BCF implementation and am having some trouble with the IDX fields. These are specified quite poorly if I may say so.

This is what the spec has to say:
```
In order to avoid this costly operation, a new IDX field can be used to explicitly define the position
which is dropped on BCF-to-VCF conversion. If not present, the implicit relative position is assumed.
If the IDX field is present in one record, it must be present also in all other dictionary-defining records.
The IDX tag is not necessary in newly created BCF files, but if present, the numbering must match
the implicit dictionary of tags.
```

The example seems to suggest the following if IDX is not present:
* The CHROM field in BCF records is always denoted by the relative offset of the *contig* entry in the header, i.e. the numbering here is independent of all non-contig header entries. [There is no "s" prefix in the spec]
* On the other hand, strings in the FILTER, INFO and FORMAT/SAMPLE fields use a joint numbering scheme. [Why? Wouldn't individual numbering allow smaller number types and independence of order between different fields? Why is this different between CHROM and the other fields?]

Now, I am really unsure what happens when IDX come into play. 
* Do they only affect FILTER, INFO and FORMAT/SAMPLE; or also CHROM?
* If the latter is the case, is counting for CHROM separate from the other fields or joint?
* Why do entries of different type but with same ID-string result in the same entry? [See below]

If I take the example from the spec and run it through bcfools convert, I get the following header:
```
##fileformat=VCFv4.3
##FILTER=<ID=PASS,Description="All filters passed",IDX=0>
##fileDate=20090805
##source=myImputationProgramV3.1
##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta
##contig=<ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species="Homo sapiens",taxonomy=x,IDX=0>
##phasing=partial
##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data",IDX=1>
##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth",IDX=2>
##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency",IDX=3>
##INFO=<ID=AA,Number=1,Type=String,Description="Ancestral Allele",IDX=4>
##INFO=<ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129",IDX=5>
##INFO=<ID=H2,Number=0,Type=Flag,Description="HapMap2 membership",IDX=6>
##FILTER=<ID=q10,Description="Quality below 10",IDX=7>
##FILTER=<ID=s50,Description="Less than 50% of samples have data",IDX=8>
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype",IDX=9>
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality",IDX=10>
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth",IDX=2>
##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality",IDX=11>
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	NA00001	NA00002	NA00003
```
As you can see, there are two entries with IDX=0 and two entries with IDX=2.

Is this behaviour intended? Obviously the numbering is not strictly increasing, so what does "the numbering must match
the implicit dictionary of tags" mean here?

Thank you for your help!

**Comment by pd3** [comment]
2. From the logic of the dictionary, a dtictionary-defining record for tag `X` is the first record that defines the string `X`, be it FILTER, INFO, or FORMAT.

1. The dictionary is implicit and unless one removes lines from the header, the IDX tags are not needed. I am curious whether htsjdk can read BCFs with removed lines and if it allows this operation. 

3. The example of IDX tag should be added to the specification

4. I don't know anything about this one

5. That is correct. The VCF specification makes the definition of the filter `PASS` optional, therefore it must be treated as implicitly present even if it does not have to explicitly appear in the header and is `IDX=0`.

6. There are two sections about it, 6.2.1 and 6.2.2. Of course the language can be improved etc. The point is there is a logic for two numbering sequences. 

7. See 2 and 6 above.

**Comment by jkbonfield** [comment]
Ah wait - I think I understand now (having noticed the mystery of why IDX=2 in this section of the top post:

```
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality",IDX=10>
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth",IDX=2>
##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality",IDX=11>
```

@pd3's example has duplicate numbering for FILTER / INFO / FORMAT fields because the `ID=` strings are duplicated too.  Ie as we have the same ID string used in FILTER and INFO then they get the same numeric value in the lookup table.  So I assume the implementation is basically a hash table insertion with auto-increment if not found, using two hashes (one for contigs and one for everything else).  If this is so, we should probably define it in exactly those terms so at least it's obvious to software developers what's going on.

Also, is that *mandatory*, or just implementation specific?  If I had a BCF implementation with separate hash tables for INFO and FORMAT, would that actually be malformed when the key strings are duplicated?  It's a subtle point and I can see how you could create an implementation without realising the mistake for a long time.  I just looked at a local BCF file with 248 ID entries and not a single one clashed, so you'd get a long way without knowing you had a bug just waiting to happen the first time someone created a reference named "GT".

**Comment by jkbonfield** [comment]
>     * re 5. That is correct. The VCF specification makes the definition of the filter `PASS` optional, therefore it must be treated as implicitly present even if it does not have to explicitly appear in the header and is `IDX=0`.

Practically speaking, what does this mean if we don't have a PASS filter?  If it's not in the header and not used in the records, does it still get allocated IDX 0 and all subsequent values start from 1?  Or is it something more generic such as any FILTER column can be used without a corresponding FILTER header tag, meaning we have to cope with scanning for filter keywords first and preallocate them in IDX?

Also the only thing mentioning that PASS doesn't need to be defined in the header (other than implicitly by numerous examples missing it out) is the BCF encoding bit.  IMO this should be explicitly stated in the VCF part of the specification.  I couldn't even find anything in the VCF bit that states the other values *must* be defined.  It explains the format of the header line, but not whether it is required.  You'd hope that's taken as read, but we know it's not the case as `##contig` for example is entirely optional (sadly).

Is it permitted to leave these lines out?

**Comment by pd3** [comment]
Even if PASS does not appear in the body, it still must be implicitly defined by the dictionary, otherwise on the first encounter of a filter one wouldn't be able to tell from the index which one we are looking at.

The VCF format allows optional definition of tags. The BCF format does not and cannot, PASS is the only exception.So I think it is alright to have the implicit index of the PASS filter mentioned in the BCF section only. I agree the language should be improved so this is stated not just as an example.

The laziness of the VCF format is another issue. I discourage the use of undefined tags because it is a problem for programs like BCFtools (or anything relying on HTSlib), but this is how VCF was originally formulated. Note that I am describing the format, not saying I like these quirks.

**Comment by h-2** [comment]
> So I assume the implementation is basically a hash table insertion with auto-increment if not found, using two hashes (one for contigs and one for everything else). If this is so, we should probably define it in exactly those terms so at least it's obvious to software developers what's going on.

Yes, that's the conclusion I came to as well. It would be really helpful if that were stated explicitly.

> Also, is that mandatory, or just implementation specific? If I had a BCF implementation with separate hash tables for INFO and FORMAT, would that actually be malformed when the key strings are duplicated? It's a subtle point and I can see how you could create an implementation without realising the mistake for a long time. I just looked at a local BCF file with 248 ID entries and not a single one clashed, so you'd get a long way without knowing you had a bug just waiting to happen the first time someone created a reference named "GT".

Yes, that is exactly what I meant above. Regardless of the answer, I have now redesigned my code to reflect the behaviour of htslib -- just to be safe.

Regarding "PASS", I am now starting the numbering for strings at 1 and always adding a PASS entry with IDX 0. I hope that's correct.
